tango/integrations/transformers/prefix_transformer.py

```diff
@@ -0,0 +1,111 @@
+import random
+from typing import Dict, Any
+import logging
+
+import torch
+from torch import nn
+from transformers import PreTrainedModel
+from transformers.modeling_outputs import CausalLMOutputWithCrossAttentions
+
+from tango.integrations.torch import Model
+
+
+logger = logging.getLogger(__name__)
+
+
+class WithPrefixEmbedding(nn.Module):
+    def __init__(
+        self, original_embedding: nn.Embedding, prefix_length: int, random_seed: int = 1940
+    ):
+        super().__init__()
+
+        self.prefix_length = prefix_length
+        self.original_embedding = original_embedding
+        self.prefix_embedding = nn.Embedding(prefix_length, self.original_embedding.embedding_dim)
+
+        # following Lester et al. 2021 in initializing using the top 5000 random vocabs
+        r = random.Random(random_seed)
+        indices = r.sample(range(5000), prefix_length)
+        with torch.no_grad():
+            self.prefix_embedding.weight.copy_(self.original_embedding.weight[indices])
+
+    def forward(self, input: torch.Tensor):
+        embedded_prefix = self.prefix_embedding(input[:, : self.prefix_length])
+        embedded_rest = self.original_embedding(input[:, self.prefix_length :])
+        return torch.cat([embedded_prefix, embedded_rest], dim=1)
+
+
+def make_prefix_transformer(model: Model, prefix_length: int) -> Model:
+    assert isinstance(model, PreTrainedModel)
+
+    model.set_input_embeddings(WithPrefixEmbedding(model.get_input_embeddings(), prefix_length))
+
+    def patch_tensor(kwargs: Dict[str, torch.Tensor], key: str, value: Any = 0) -> None:
+        try:
+            t = kwargs[key]
+        except KeyError:
+            return
+        zero_prefix = torch.full(
+            (t.size(0), prefix_length) + t.shape[2:], value, dtype=t.dtype, device=t.device
+        )
+        kwargs[key] = torch.cat([zero_prefix, t], dim=1)
+
+    def patch_tensor_with_indices(kwargs: Dict[str, torch.Tensor], key: str) -> None:
+        try:
+            t = kwargs[key]
+        except KeyError:
+            return
+        kwargs[key] = torch.cat(
+            [
+                torch.arange(0, prefix_length, dtype=t.dtype)
+                .unsqueeze(0)
+                .expand(t.size(0), prefix_length),
+                t + prefix_length,
+            ],
+            dim=1,
+        )
+
+    # Because PyTorch hooks don't support kwargs, we monkey patch the forward method 🙈
+    old_forward = model.forward
+
+    def new_forward(*args, **kwargs):
```

Comment:
What made me turn away from monkeypatching in my code is when I noticed that it doesn't need/have a `self`, so there might be some fundamental differences between the old vs. new forward. If I were two years younger I probably would have voted for monkeypatching but the older me is less adventurous and worry more about safety. Go ahead if you're confident that this is safe, but at least I would suggest some sort of assertion to check the forward has not been monkeypatched (because if it had, the logic would be incorrect).

Reply:
I am a little uneasy about it, but I think it beats the alternatives. At the very least I want to see where it goes and where it falls down, if it does. Also, apparently there is movement on the PyTorch side to allow `kwargs` in hooks. When that comes true, we can do this properly.

As for your specific concern, this will work fine even if `forward()` has already been monkey patched before.

Reply:
Will it? Wouldn't be the patching happen multiple times at each recursion level?

Reply:
If the thing you pass into this function was monkey patched before, then `old_forward` ends up being the first level of monkey patching, and it will get called when we go one level down.

`old_forward` becomes part of the closure of `new_forward`. That's how the chain of forward methods is maintained.

Reply:
Right. So it would be like this, no?
```
forward:  # monkey patch lvl 1
  patch_tensor
  forward:  # monkey patch lvl 2  
    patch_tensor
    forward  # original
```

Reply:
Yes

Reply:
So the tensors will be patched twice

Reply:
The inputs and outputs should be patched twice. That is correct.

What won't work is that you can't call `set_input_embeddings()` twice with the way I have it here, because `_WithPromptEmbedding` reaches into the original embedding's internals.

Reply:
Oh, what I was worried about is the method being unintentionally called twice. I can't think of a case where it is intentionally called twice.

Reply:
c8d1b86 should make it possible to stack two prompt-enabled transformers on top of each other.

I think it's important that we ensure this pattern works for other forms as well. What if we implement adapters the same way, and we want to run both at the same time? The whole point of trying for this "looks like a normal huggingface transformer" approach is that it should be easy to combine with other components that do the same thing.